lib/ipam/ipam.go

```diff
@@ -119,7 +125,24 @@ func (c ipamClient) AutoAssign(ctx context.Context, args AutoAssignArgs) ([]net.
 		v6list, err = c.autoAssign(ctx, args.Num6, args.HandleID, args.Attrs, args.IPv6Pools, 6, hostname, args.MaxBlocksPerHost, args.HostReservedAttrIPv6s)
 		if err != nil {
 			log.Errorf("Error assigning IPV6 addresses: %v", err)
-			return v4list, v6list, err
+		}
+
+		if len(v6list) != args.Num6 {
+			if len(v4list) != 0 {
+				// Release any v4 IPs that were assigned when we were expecting to assign
+				// both v4 and v6 IPs and v6 IP assignment did not complete.
+				// This should only occur in dual stack environments.
+				// This prevents v4 IPs from being assigned when v6 assignment fails.
+				v4IPs := []net.IP{}
+				for _, v4 := range v4list {
+					v4IPs = append(v4IPs, *net.ParseIP(v4.IP.String()))
+				}
+				_, releaseErr := c.ReleaseIPs(ctx, v4IPs)
+				if releaseErr != nil {
+					log.Errorf("Error releasing IPv4 addresses %+v on IPv6 address assignment failure: %s", v4IPs, err)
```

Comment:
I think if we hit this line, we're going to leak the v4 addresses - we'll tell the caller that no v4 addresses were allocated, but actually some were.